File: src/test/java/com/polygloat/controllers/UserControllerTest.java
package com.polygloat.controllers;

import com.polygloat.Assertions.StandardValidationMessageAssert;
import com.polygloat.dtos.request.UserUpdateRequestDTO;
import com.polygloat.model.UserAccount;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
import org.springframework.test.web.servlet.MvcResult;
import org.testng.annotations.Test;

import java.util.Optional;

import static com.polygloat.Assertions.Assertions.assertThat;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

public class UserControllerTest extends SignedInControllerTest implements ITest {

    @Test
    void updateUser() throws Exception {
        UserUpdateRequestDTO requestDTO = UserUpdateRequestDTO.builder().email("ben@ben.aa").password("super new password").name("Ben's new name").build();

        MvcResult mvcResult = performPost("/api/user", requestDTO).andExpect(status().isOk()).andReturn();

        Optional<UserAccount> fromDb = userAccountService.getByUserName(requestDTO.getEmail());
        assertThat(fromDb).isNotEmpty();

        BCryptPasswordEncoder bCryptPasswordEncoder = new BCryptPasswordEncoder();
        assertThat(bCryptPasswordEncoder.matches(requestDTO.getPassword(), fromDb.get().getPassword())).describedAs("Password is changed").isTrue();

        assertThat(fromDb.get().getName()).isEqualTo(requestDTO.getName());
    }

    @Test
    void updateUserValidation() throws Exception {
        UserUpdateRequestDTO requestDTO = UserUpdateRequestDTO.builder().email("ben@ben.aa").password("").name("").build();

        MvcResult mvcResult = performPost("/api/user", requestDTO).andExpect(status().isBadRequest()).andReturn();

        StandardValidationMessageAssert standardValidation = assertThat(mvcResult).error().isStandardValidation();
        standardValidation.onField("password");
        standardValidation.onField("name");

        requestDTO = UserUpdateRequestDTO.builder().email("ben@ben.aa").password("aksjhd  dasdsa").name("a").build();
        dbPopulator.createUser(requestDTO.getEmail());
        mvcResult = performPost("/api/user", requestDTO).andExpect(status().isBadRequest()).andReturn();
        assertThat(mvcResult).error().isCustomValidation().hasMessage("username_already_exists");
    }

}

File: src/test/java/io/tolgee/controllers/UserControllerTest.kt
package io.tolgee.controllers

import io.tolgee.ITest
import io.tolgee.dtos.request.UserUpdateRequestDTO
import org.assertj.core.api.Assertions
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder
import org.springframework.test.web.servlet.result.MockMvcResultMatchers
import org.testng.annotations.Test

class UserControllerTest : SignedInControllerTest(), ITest {
    @Test
    fun updateUser() {
        val requestDTO = UserUpdateRequestDTO(
                email = "ben@ben.aa",
                password = "super new password",
                name = "Ben's new name"
        )
        performAuthPost("/api/user", requestDTO).andExpect(MockMvcResultMatchers.status().isOk)
        val fromDb = userAccountService.getByUserName(requestDTO.email)
        Assertions.assertThat(fromDb).isNotEmpty
        val bCryptPasswordEncoder = BCryptPasswordEncoder()
        Assertions.assertThat(bCryptPasswordEncoder.matches(requestDTO.password, fromDb.get().password)).describedAs("Password is changed").isTrue
        Assertions.assertThat(fromDb.get().name).isEqualTo(requestDTO.name)
    }

    @Test
    fun updateUserValidation() {
        var requestDTO = UserUpdateRequestDTO(
                email = "ben@ben.aa",
                password = "",
                name = ""
        )
        var mvcResult = performAuthPost("/api/user", requestDTO).andExpect(MockMvcResultMatchers.status().isBadRequest).andReturn()
        val standardValidation = io.tolgee.assertions.Assertions.assertThat(mvcResult).error().isStandardValidation
        standardValidation.onField("password")
        standardValidation.onField("name")

        requestDTO = UserUpdateRequestDTO(
                email = "ben@ben.aa",
                password = "aksjhd  dasdsa",
                name = "a"
        )
        dbPopulator.createUserIfNotExists(requestDTO.email!!)
        mvcResult = performAuthPost("/api/user", requestDTO).andExpect(MockMvcResultMatchers.status().isBadRequest).andReturn()
        io.tolgee.assertions.Assertions.assertThat(mvcResult).error().isCustomValidation.hasMessage("username_already_exists")
    }
}
